modules/configuration/src/main/java/org/apache/ignite/configuration/internal/validation/ValidationUtil.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.configuration.internal.validation;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import org.apache.ignite.configuration.RootKey;
+import org.apache.ignite.configuration.internal.RootsNode;
+import org.apache.ignite.configuration.internal.util.AnyNodeConfigurationVisitor;
+import org.apache.ignite.configuration.internal.util.KeysTrackingConfigurationVisitor;
+import org.apache.ignite.configuration.tree.InnerNode;
+import org.apache.ignite.configuration.validation.ValidationIssue;
+import org.apache.ignite.configuration.validation.Validator;
+
+import static java.util.Collections.emptySet;
+import static org.apache.ignite.configuration.internal.util.ConfigurationUtil.appendKey;
+
+/** */
+public class ValidationUtil {
+    /**
+     * Validate configuration changes.
+     *
+     * @param oldRoots Old known roots.
+     * @param newRoots New roots.
+     * @param otherRoots Provider for arbitrary roots that might not be accociated with the same storage.
+     * @param memberAnnotationsCache Mutable map that contains annotations associated with corresponding member keys.
+     * @param validators Current validators map to look into.
+     * @return List of validation results.
+     */
+    public static List<ValidationIssue> validate(
+        RootsNode oldRoots,
+        RootsNode newRoots,
+        Function<RootKey<?, ?>, InnerNode> otherRoots,
+        Map<MemberKey, Annotation[]> memberAnnotationsCache,
```

Comment:
I personally don't like `MemberKey` name as well, it is not informative. Ii would be great if we can come up with something more intuitive.

But it is definitely not a stopper for this PR.

Reply:
I don't like it either.